### PR TITLE
Add `LazyKernel` for lazily-compiled custom `cupy` kernels

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

Add a wrapper around `cupy.RawKernel` that lazily compiles them on first invocation.

This prevents CuPy from allocating memory unnecessarily during module init.

### Types of change

Refactor

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by shadeMe** [comment]
@explosion-bot please test_slow_gpu

File: thinc/backends/_custom_kernels.py [hunk 1]
```diff
@@ -62,74 +62,106 @@
...
+
+    def _compile_kernel(self):
+        if self._compiled:
+            return
+
+        if self._compile_callback is not None:
+            self._kernel = self._compile_callback()
+        elif KERNELS is not None:
+            self._kernel = KERNELS.get_function(self.name)
+        self._compiled = True
```
Comment (danieldk) [inline thinc/backends/_custom_kernels.py:106]: We always have `KERNELS` I think? So both code paths should return something (or raise an exception). I am wondering if we can't do without `_compiled` and signify that `self._kernel is None` means that the kernel is not yet compiled. But I might definitely be overlooking something here.

**Review by danieldk — COMMENTED** [review]
Nice improvement! I wonder if we lost lazy compilation somewhere along the way or I just misremember (I seem to remember that I could see kernels getting added to the compilation cache as they were used).

Two small comments/questions.

File: thinc/backends/_custom_kernels.py (hunk 1 above)
Comment (shadeMe) [reply thinc/backends/_custom_kernels.py:106]: Looking at the initialization code, it seems like `KERNELS` is `None` when `has_cupy_gpu` is `False`. But it is true that `CupyOps` won't ever be called in that case, so I think we can get rid of the flag.

File: thinc/backends/_custom_kernels.py [hunk 2]
```diff
@@ -62,74 +62,106 @@
...
-backprop_dish_kernel_float = _get_kernel("backprop_dish<float>")
-backprop_gelu_kernel_double = _get_kernel("backprop_gelu<double>")
-backprop_gelu_kernel_float = _get_kernel("backprop_gelu<float>")
-backprop_hard_swish_kernel_double = _get_kernel("backprop_hard_swish<double>")
-backprop_hard_swish_kernel_float = _get_kernel("backprop_hard_swish<float>")
-backprop_hard_swish_mobilenet_kernel_double = _get_kernel(
+    return cupy.RawKernel((PWD / "_murmur3.cu").read_text(encoding="utf8"), "hash_data")
+
+
+clipped_linear_kernel_float = LazyKernel("clipped_linear<float>")
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply thinc/backends/_custom_kernels.py:111]: Done.

**Comment by shadeMe** [comment]
@explosion-bot please test_slow_gpu

**Review by danieldk — APPROVED** [review]

**Pull request merged**